kernel/src/fs/sysfs/inode.rs

```diff
@@ -0,0 +1,827 @@
+                let special_iter = ThisAndParentDentryIter::new(self, min_ino);
+                SysFsDentryIter::new(attr_iter, node_iter, special_iter)
+            }
+            InnerNode::Attr(_, _) | InnerNode::Symlink(_) => {
+                // Non-directories: return empty iterators for attrs and nodes
+                let attr_iter = AttrDentryIter::new(Vec::new(), self.ino(), min_ino);
+                let node_iter = NodeDentryIter::new(Vec::new(), min_ino);
+                // . and .. might not make sense here, but keep for consistency or adjust readdir_at logic
+                let special_iter = ThisAndParentDentryIter::new(self, min_ino);
+                SysFsDentryIter::new(attr_iter, node_iter, special_iter)
+            }
+        }
+    }
+}
+
+impl Inode for SysFsInode {
+    fn type_(&self) -> InodeType {
+        self.metadata.type_
+    }
+
+    fn metadata(&self) -> Metadata {
+        self.metadata
+    }
+
+    fn ino(&self) -> u64 {
+        self.metadata.ino
+    }
+
+    fn mode(&self) -> Result<InodeMode> {
+        Ok(*self.mode.read())
+    }
+
+    fn set_mode(&self, mode: InodeMode) -> Result<()> {
+        *self.mode.write() = mode;
+        Ok(())
+    }
+
+    fn size(&self) -> usize {
+        self.metadata.size
+    }
+
+    fn resize(&self, _new_size: usize) -> Result<()> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn atime(&self) -> Duration {
+        self.metadata.atime
+    }
+    fn set_atime(&self, _time: Duration) {}
+
+    fn mtime(&self) -> Duration {
+        self.metadata.mtime
+    }
+    fn set_mtime(&self, _time: Duration) {}
+
+    fn ctime(&self) -> Duration {
+        self.metadata.ctime
+    }
+    fn set_ctime(&self, _time: Duration) {}
+
+    fn owner(&self) -> Result<Uid> {
+        Ok(self.metadata.uid)
+    }
+    fn set_owner(&self, _uid: Uid) -> Result<()> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn group(&self) -> Result<Gid> {
+        Ok(self.metadata.gid)
+    }
+    fn set_group(&self, _gid: Gid) -> Result<()> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn fs(&self) -> Arc<dyn FileSystem> {
+        crate::fs::sysfs::singleton().clone()
+    }
+
+    fn page_cache(&self) -> Option<crate::vm::vmo::Vmo<aster_rights::Full>> {
+        None
+    }
+
+    fn read_at(&self, offset: usize, buf: &mut VmWriter) -> Result<usize> {
+        self.read_direct_at(offset, buf)
+    }
+
+    fn read_direct_at(&self, _offset: usize, buf: &mut VmWriter) -> Result<usize> {
+        // TODO: it is unclear whether we should simply ignore the offset
+        // or report errors if it is non-zero.
+
+        let InnerNode::Attr(attr, leaf) = &self.inner_node else {
+            return Err(Error::new(Errno::EINVAL));
+        };
+
+        // TODO: check read permission
+        match leaf.read_attr(attr.name(), buf) {
+            Ok(len) => Ok(len),
+            Err(_) => {
+                return_errno_with_message!(Errno::EIO, "failed to read attribute");
```

Comment:
Again, implement the `From` trait for the error type of systree so that you don't have to convert error types manually at each location.